src/main/java/htsjdk/samtools/CigarOperator.java

```diff
@@ -89,114 +117,116 @@
+    }
+
+    public static CigarOperator fromChar(final int ch) {
+       if ((ch & ~0x7F) == 0) {
+           final CigarOperator result = byte2Value[ch & 0x7F];
+           if (result == null) {
+               throw new IllegalArgumentException("Unrecognized CigarOperator: " + ch);
+           }
+           return result;
+       } else {
+           throw new IllegalArgumentException("Unrecognized CigarOperator: " + ch);
+       }
     }
 
     /**
      * @param i CIGAR operator in binary form as appears in a BAMRecord.
      * @return CigarOperator enum value corresponding to the given int value.
+     * @deprecated use {@link #fromBamCode(int)}.
      */
+    @Deprecated
     public static CigarOperator binaryToEnum(final int i) {
-        switch(i) {
-            case OP_M:
-                return M;
-            case OP_I:
-                return I;
-            case OP_D:
-                return D;
-            case OP_N:
-                return N;
-            case OP_S:
-                return S;
-            case OP_H:
-                return H;
-            case OP_P:
-                return P;
-            case OP_EQ:
-                return EQ;
-            case OP_X:
-                return X;
-            default:
-                throw new IllegalArgumentException("Unrecognized CigarOperator: " + i);
+        return fromBamCode(i);
+    }
+
+    /**
+     * Looks-up the operator given its BAM code.
+     * <p>
+     *     It is guaranteed that {@code CigarOperator.fromBamCode(op.bamCode()) == op}.
+     * </p>
+     *
+     * @param bamCode the query BAM code.
+     * @return never {@code null}.
+     * @throws IllegalArgumentException if the input BAM code is not valid.
+     * @see #bamCode
+     */
+    public static CigarOperator fromBamCode(final int bamCode) {
+        if (bamCode >= 0 && bamCode < bamCode2Value.length) {
+            return bamCode2Value[bamCode];
+        } else {
+            throw new IllegalArgumentException("Unrecognized CigarOperator: " + bamCode);
         }
     }
 
     /**
      *
      * @param e CigarOperator enum value.
      * @return CIGAR operator corresponding to the enum value in binary form as appears in a BAMRecord.
+     * @deprecated use {@link #bamCode()} instead.
      */
+    @Deprecated
     public static int enumToBinary(final CigarOperator e) {
-        switch(e) {
-            case M:
-                return OP_M;
-            case I:
-                return OP_I;
-            case D:
-                return OP_D;
-            case N:
-                return OP_N;
-            case S:
-                return OP_S;
-            case H:
-                return OP_H;
-            case P:
-                return OP_P;
-            case EQ:
-                return OP_EQ;
-            case X:
-                return OP_X;
-            default:
-                throw new IllegalArgumentException("Unrecognized CigarOperator: " + e);
-        }
+        return e.ordinal();
```

Comment:
I don't think that it is safe to use ordinal() for encoding and decoding. as it means that two implicit decision (the order of the enum definition, and the way java converts an enum class into ordinals) affects how the cigar is encoded and decoded. I think that it is much preferable to have it written explicitly. 

@tfenne @lbergelson opinions?

Reply:
I added unit test to check that the transcription into bam-codes is correct. So a change in order of the enum constant would result in a test failure. 
However I'm happy to change it to explicitly indicate the code as a additional constructor argument.

Reply:
I think it's unlikely that this will be a problem in practice, but typically it's considered bad practice to rely on the ordinal value in java because it's brittle if you change the ordering or add new enum values.  I would say keep it as a constant in the constructor.